crates/stages/src/stages/hashing_storage.rs

```diff
@@ -0,0 +1,491 @@
+use crate::{
+    db::Transaction, ExecInput, ExecOutput, Stage, StageError, StageId, UnwindInput, UnwindOutput,
+};
+use reth_db::{
+    cursor::{DbCursorRO, DbCursorRW, DbDupCursorRO, DbDupCursorRW},
+    database::Database,
+    models::TransitionIdAddress,
+    tables,
+    transaction::{DbTx, DbTxMut},
+};
+use reth_primitives::{keccak256, Address, StorageEntry, H160, H256, U256};
+use std::{collections::BTreeMap, fmt::Debug};
+use tracing::*;
+
+const STORAGE_HASHING: StageId = StageId("StorageHashingStage");
+
+/// Storage hashing stage hashes plain storage.
+/// This is preparation before generating intermediate hashes and calculating Merkle tree root.
+#[derive(Debug)]
+pub struct StorageHashingStage {
+    /// The threshold for switching from incremental hashing
+    /// of changes to whole storage hashing. Num of transitions.
+    pub clean_threshold: u64,
+    /// The size of inserted items after which the control
+    /// flow will be returned to the pipeline for commit
+    pub commit_threshold: u64,
+}
+
+#[async_trait::async_trait]
+impl<DB: Database> Stage<DB> for StorageHashingStage {
+    /// Return the id of the stage
+    fn id(&self) -> StageId {
+        STORAGE_HASHING
+    }
+
+    /// Execute the stage.
+    async fn execute(
+        &mut self,
+        tx: &mut Transaction<'_, DB>,
+        input: ExecInput,
+    ) -> Result<ExecOutput, StageError> {
+        let stage_progress = input.stage_progress.unwrap_or_default();
+        let previous_stage_progress = input.previous_stage_progress();
+
+        // read storage changeset, merge it into one changeset and calculate storage hashes.
+        let from_transition = tx.get_block_transition(stage_progress)? + 1;
+        let to_transition = tx.get_block_transition(previous_stage_progress)? + 1;
+
+        // if there are more blocks then threshold it is faster to go over Plain state and hash all
+        // account otherwise take changesets aggregate the sets and apply hashing to
+        // AccountHashing table
+        if to_transition - from_transition > self.clean_threshold {
+            tx.clear::<tables::HashedStorage>()?;
+            tx.commit()?;
```

Comment:
How does Erigon do this? Do they drop the table every time?

Reply:
Not every time, but if a lot of blocks are executed at once. It made sense for me as applying every changetset would take a lot of time. Erigon does in intermediatehash stage: https://github.com/ledgerwatch/erigon/blob/937fb2c4d95b0640904b7dd715c1736e7d0a5ede/eth/stagedsync/stage_interhashes.go#L149-L150